### PR TITLE
TRUNK-5053 : Refactor ProgramWorkflowServiceImpl.saveProgram

### PR DESCRIPTION
**Pull request by JudeNiroshan**
Refactor ProgramWorkflowServiceImpl.saveProgram

## Issue I worked on

see https://issues.openmrs.org/browse/TRUNK-5053

## Checklist: I completed these to help reviewers :)

- [x] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

  No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [x] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: api/src/main/java/org/openmrs/api/impl/ProgramWorkflowServiceImpl.java [hunk 1]
```diff
@@ -69,38 +69,57 @@ public void setProgramWorkflowDAO(ProgramWorkflowDAO dao) {
...
-				}
+				ensureProgramWorkflowIsSet(state, workflow);
 			}
 		}
 		return dao.saveProgram(program);
 	}
+
+	/**
+	 * Helper method to ensure the program is set to the given workflow by doing the validations
+	 * @param workflow
```
Comment (teleivo) [inline api/src/main/java/org/openmrs/api/impl/ProgramWorkflowServiceImpl.java]: @JudeNiroshan please remove your javadocs. They dont add value/information. The methods are short, dont do much, and have meaningful names. Its easy for the reader to see what they do. And most importantly these are private methods. If they were public then of course you need to add proper javadocs (but then you would also need to fill in the parameter doc as codacy told you)

File: api/src/test/java/org/openmrs/api/ProgramWorkflowServiceUnitTest.java [hunk 2]
```diff
@@ -14,12 +14,19 @@
...
+import org.junit.rules.ExpectedException;
 import org.mockito.Mockito;
+import org.openmrs.Concept;
 import org.openmrs.Program;
+import org.openmrs.ProgramWorkflow;
+import org.openmrs.ProgramWorkflowState;
 import org.openmrs.api.db.ProgramWorkflowDAO;
 import org.openmrs.api.impl.ProgramWorkflowServiceImpl;
 
+import static org.junit.Assert.assertTrue;
```
Comment (teleivo) [inline api/src/test/java/org/openmrs/api/ProgramWorkflowServiceUnitTest.java]: @JudeNiroshan please configure your IDE as per https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE
especially the imports (static ones should be on top).

Comment (JudeNiroshan) [inline api/src/test/java/org/openmrs/api/ProgramWorkflowServiceUnitTest.java]: @teleivo  I have actually configured accordingly. I'm using IDEA

Comment (teleivo) [inline api/src/test/java/org/openmrs/api/ProgramWorkflowServiceUnitTest.java]: https://wiki.openmrs.org/display/docs/Developer+How-To+Setup+And+Use+IntelliJ#DeveloperHow-ToSetupAndUseIntelliJ-JavaImports

static imports are first, check again

Comment (JudeNiroshan) [inline api/src/test/java/org/openmrs/api/ProgramWorkflowServiceUnitTest.java]: @teleivo My apologies. I missed it. Currently I'm writing the test to get 100% coverage. Once I'm done with that, I will commit them.

**Comment by teleivo** [comment]
@JudeNiroshan I am closing this PR for now since we have many PRs open and it becomes messy for us reviewers :) Please fell free to create a new PR when you have addressed my comments. Thank you very much for your contribution!

**Pull request closed without merging**